### ISSUE TITLE
Success Criterion 4.1.1 Parsing

### ISSUE THREAD

**Issue by maryjom** [w3c/wcag2ict#241]
#### Success Criterion 4.1.1 Parsing

[Success Criterion 4.1.1 Parsing](https://www.w3.org/TR/WCAG22/#parsing):

>NOTE
>This criterion was originally adopted to address problems that assistive technology had directly parsing HTML. Assistive technology no longer has any need to directly parse HTML. Consequently, these problems either no longer exist or are addressed by other criteria. This criterion no longer has utility and is removed.

##### Guidance on Applying Success Criterion 4.1.1 to Non-Web Documents and Software

**WCAG 2.2 guidance:**

This success criterion has been made obsolete and was removed from WCAG 2.2 for Web content. Therefore, it should also be made obsolete and removed as a requirement for non-web documents and software. Assistive technology does not parse non-web documents or software to obtain accessibility information, but instead relies on the user agent or platform software to provide that information through the accessibility APIs. 

**WCAG 2.0 and 2.1 guidance:**

WCAG 2.0 and 2.1 are incorporated, either directly or by reference, into other standards. It follows that the application of 4.1.1 Parsing to non-web documents and software should follow the new guidance provided in the [WCAG 2.0 Editorial Errata](https://www.w3.org/WAI/WCAG20/errata/#editorial) and the [WCAG 2.1 Editorial Errata](https://www.w3.org/WAI/WCAG21/errata/#editorial) which state the following:
> "This Success Criterion should be considered as always satisfied for any content using HTML or XML."

For non-web documents and software, this statement applies directly as written, replacing “HTML or XML” with “markup languages in such a way that the markup is separately exposed and available to [assistive technologies](https://www.w3.org/TR/wcag2ict-22/#dfn-assistive-technology) and accessibility features of software or to a user-selectable [user agent](https://www.w3.org/TR/wcag2ict-22/#user-agent)”.

With these substitutions, it would read:

> "This Success Criterion should be considered as always satisfied for any content using **[markup languages in such a way that the markup is separately exposed and available to [assistive technologies](https://www.w3.org/TR/wcag2ict-22/#dfn-assistive-technology) and accessibility features of software or to a user-selectable [user agent](https://www.w3.org/TR/wcag2ict-22/#user-agent)]**."

**NOTE:**
When conformance to WCAG 2.0 or 2.1 is required:

- If non-web documents or software are implemented using markup language, this Success Criterion is automatically satisfied. 
- If non-web documents or software are not implemented using a markup language, this Success Criterion would not be applicable.

---
In the section on SC problematic for closed functionality, modify the existing bullet as follows:
#### Success Criteria Problematic for Closed Functionality

**4.1.1 Parsing** —The intent of this success criterion was focused on using correct HTML syntax in content developed using markup languages so that different user agents or assistive technologies would yield the same result; since WCAG 2.2 has removed this success criterion, it should not be applied to closed functionality software.

**Comment by JAWS-test** [comment]
> This success criterion has been made obsolete and was removed from WCAG 2.2 for Web content. Therefore, it should also be made obsolete and removed as a requirement for non-web documents and software. 

If I understand correctly, the W3C has decided to remove 4.1.1, 
* because the HTML specification now clearly states how syntax errors are to be corrected.
* because the AT does not analyse the source code, but accesses the API.

As soon as one of the two conditions does not apply, 4.1.1 would continue to be relevant. 

4.1.1 states:

> This Success Criterion should be considered as always satisfied for any content using HTML or XML.

For me, this implies that 4.1.1 can still be relevant for other markup languages. Example: If there are no rules in the SVG or PDF specification on how to deal with duplicate IDs, then depending on the user agent, the output to the API can be different for duplicate IDs and 4.1.1 is important.

I suggest investigating whether all markup languages for software and documents really define rules on how to deal with syntax errors and whether in each case the AT accesses the API and not the source code.

**Comment by maryjom** [comment]
@JAWS-test There is no WCAG requirement for parsing in WCAG 2.2, which means WCAG2ICT has nothing to interpret for non-web technologies.  If there is actually a need for a similar requirement for non-web technologies, then other standards, like the PDF standard would have to have a requirement for proper beginning/end tags and duplicate IDs. If it is super important to have such a requirement for accessibility of non-web documents, then the EN 301 549 and 508 standards should take that up.

**Comment by maryjom** [comment]
## Background on the new proposals
Based on feedback here, in the survey and in the 26 October meeting discussion, we decided that the guidance we put in this WCAG2ICT document (or at the very least in these SC by SC guidance sections) is **ONLY in the context of WCAG 2.2**.  We'll have to figure out what to do with WCAG 2.0 and 2.1 and where/how that would be documented after we discuss with the AG WG leadership. 

So for the upcoming survey there will be 3 options for the WCAG2ICT guidance in the context of WCAG 2.2.

1. State that since this was removed for Web content, it should be removed for non-web (original proposed text). This would also require vetting with AT developers that ATs don't parse source documents for accessibility information and instead rely upon user agents to do that part.
2. State that since there is no longer a WCAG criterion to interpret, the guidance has been removed. 
3. Could further state, in addition to point 2, a note that if there is still a need for a parsing requirement in a non-web context that it would be up to other standards bodies to make that determination. 

------

#### [Success Criterion 4.1.1 Parsing](https://www.w3.org/TR/WCAG22/#parsing)

>NOTE
>This criterion was originally adopted to address problems that assistive technology had directly parsing HTML. Assistive technology no longer has any need to directly parse HTML. Consequently, these problems either no longer exist or are addressed by other criteria. This criterion no longer has utility and is removed.

##### Applying Success Criterion 4.1.1 Parsing to Non-Web Documents and Software

NOTE: 

**Option 1**
This success criterion has been made obsolete and was removed from WCAG 2.2 for Web content. Therefore, it should also be made obsolete and removed as a requirement for non-web documents and software. Assistive technology does not parse non-web documents or software to obtain accessibility information, but instead relies on the user agent or platform software to provide that information through the accessibility APIs. 

**Option 2**
WCAG 2.2 has made this success criterion obsolete and removed it as a requirement in the standard. Therefore, the interpretation of this success criterion for non-web documents and software has been removed.

**Option 3**
WCAG 2.2 has made this success criterion obsolete and removed it as a requirement in the standard. Therefore, the interpretation of this success criterion for non-web documents and software has been removed. 

NOTE: The task force does not know whether assistive technologies directly parse any non-web markup languages to obtain accessibility information. Another unknown is whether user agents for rendering content using non-web markup languages have standard methods for correcting or addressing poor markup that doesn't meet the markup specification. In such cases, other standards bodies would need to investigate whether requirements are needed to ensure the accessibility of content produced in those non-web markup languages.

---
### SC Problematic for Closed Functionality
In this section, remove the existing bullet for 4.1.1 Parsing.

**Comment by awkawk** [comment]
Regarding PDF, it is important to point out first that there is no markup as PDF is not a markup language. This is why in the definition for Programmatically Determined (https://www.w3.org/TR/WCAG21/#dfn-programmatically-determinable) there are two options:

- Determined in a markup language from elements and attributes that are accessed directly by commonly available assistive technology.
- Determined from technology-specific data structures in a non-markup language and exposed to assistive technology via an accessibility API that is supported by commonly available assistive technology.

The second is applies to PDF. There is an accessibility API in Adobe Reader that acts as an intermediate API when then maps the information to the appropriate API for the platform, and that is what the assistive technologies interface with.

4.1.1 has never applied to PDF or other non-markup content (e.g., Silverlight, Flash, video, audio).

**Comment by maryjom** [comment]
Thanks for the quick response, Andrew. I didn’t know much about the underlying technology of PDF, but I did suspect that the accessibility information is actually presented to AT via the accessibility API in the end.  So the conclusion is that 4.1.1 Parsing should not be applied to non-web software or documents (and probably shouldn’t have been before it was made obsolete and deleted in WCAG 2.2.

Best regards,

Mary Jo Mueller
IBM Accessibility Standards Program Manager


From: Andrew Kirkpatrick ***@***.***>
Date: Monday, October 30, 2023 at 3:43 PM
To: w3c/wcag2ict ***@***.***>
Cc: Mary Jo Mueller ***@***.***>, Assign ***@***.***>
Subject: [EXTERNAL] Re: [w3c/wcag2ict] Success Criterion 4.1.1 Parsing (Issue #241)
Regarding PDF, it is important to point out first that there is no markup as PDF is not a markup language. This is why in the definition for Programmatically Determined (https: //www. w3. org/TR/WCAG21/#dfn-programmatically-determinable) there
ZjQcmQRYFpfptBannerStart
This Message Is From an External Sender
This message came from outside your organization.
    Report Suspicious  <https://us-phishalarm-ewt.proofpoint.com/EWT/v1/PjiDSg!2g-pItQcolPwkMXw8unGZfXnzPxzAzJ54Xqpc6N60CIQeFg3QK70B_jbf_xLfD1Jr1l4BJw9Th8WjyRKj8eUAgGUlznxGYk$>   ‌
ZjQcmQRYFpfptBannerEnd

Regarding PDF, it is important to point out first that there is no markup as PDF is not a markup language. This is why in the definition for Programmatically Determined (https://www.w3.org/TR/WCAG21/#dfn-programmatically-determinable<https://www.w3.org/TR/WCAG21/#dfn-programmatically-determinable>) there are two options:

  *   Determined in a markup language from elements and attributes that are accessed directly by commonly available assistive technology.
  *   Determined from technology-specific data structures in a non-markup language and exposed to assistive technology via an accessibility API that is supported by commonly available assistive technology.

The second is applies to PDF. There is an accessibility API in Adobe Reader that acts as an intermediate API when then maps the information to the appropriate API for the platform, and that is what the assistive technologies interface with.

4.1.1 has never applied to PDF or other non-markup content (e.g., Silverlight, Flash, video, audio).

**Comment by maryjom** [comment]
So this doesn't get lost...Gregg had proposed some text for the instances of WCAG 2.0 and 2.1 (which will be handled outside of the document we're working on for interpretation of WCAG 2.2.) I've done some editing and omitted WCAG 2.2. I also removed the history of the development of the WCAG language, as the note in WCAG 2.1 and errata for WCAG 2.0 handle that aspect.
**Gregg's proposal from the 4.1.1 Parsing survey**
>4.1.1, which requires syntax of web pages not have unparseable content without repair, was required in the past (10 years ago) because AT parsed content itself and was not as good at parsing as browsers. AT no longer parses on its own and this problem no longer exists for web content. It is no longer needed in 2.2 and is also no longer needed when following 2.1 or 2.0 -- since the problem from the past no longer exists with today's AT. For software and non-web documents, there is no known problem with AT anymore either so 4.1.1 can be treated in the same way as it is in 2.2 and as it is recommended to be treated in 2.1 and 2.0 - and that is to be ignored. It does not have any effect on the accessibility of web content anymore and is not known to have any effect on software or non-web content accessibility either - so should no longer be a requirement for accessibility.

**Mary Jo's alternate proposed alternative language:** 
(for WCAG 2.0 and 2.1)
As in Web content, 4.1.1 Parsing is not known to have any effect on the accessibility of non-web documents or software. There are no known examples of non-web documents or software that would have an issue such as those covered by 4.1.1 Parsing. Modern assistive technology does not parse document or software markdown languages for accessibility information. User agents and platforms used to render non-web content and software use platform accessibility APIs to present accessibility information to AT. Therefore, 4.1.1 Parsing should no longer be a requirement for accessibility.

**Comment by mraccess77** [comment]
HTML content in an app like an Electron app (e.g. Teams, Slack) is non-web content because web content is very scoped to a user agent and a url.  But this content is created with a markup language.  So I would prefer we be very careful in our wording.  

On separate note, In web content - I use assistive browser extensions like read aloud tools that access the DOM and don't use the accessibility API - while they are not parsers - they are using the DOM structure to interpret and red the content.  So we should be very careful to not assume that AT doesn't use the DOM - while screen readers like NVDA and JAWS may not there are many ATs that are extensions that do use the DOM and it's structure.

**Comment by JAWS-test** [comment]
> I agree that 4.1.1 has never been applicable to PDF. That said, my understanding is that PDF evolved from PostScript

I think that PDF documents can be checked for 4.1.1. A PDF from Postscript is not accessible at all. The accessibility in the PDF results from the PDF tags that can be added to the Postscript file. The PDF tags work very similarly to HTML tags, most of them even have the same name (e.g. P for paragraph, H1 for heading). As far as I know, the [PAC](https://pdfua.foundation/en/) checks the correct syntax of the tags according to WCAG 4.1.1. The [PDF Accessibility Checklist](https://abledocs.com/da/pdf-accessibility-checkliste/) also explains the 4.1.1 check for PDFs

**Comment by maryjom** [comment]
@JAWS-test See [Andrew Kirkpatrick's comment](https://github.com/w3c/wcag2ict/issues/241#issuecomment-1785921498) on the related issue when the TF was working drafting guidance on 4.1.1 Parsing. According to this post, PDF is not a markup lanugage.

**Comment by maryjom** [comment]
Closing this issue, as the SC is up-to-date and is under review of the AG WG for the 2nd time. The TF resolved concerns on the first AG WG review and approved merging in those changes on 16 May. The new AG WG review issue #364 supersedes this and if there are any further concerns raised, they'll be handled with that issue.
